### PR TITLE
Perf: add support for launch ID

### PR DESCRIPTION
**Pull request by EdmondChuiHW**
# Summary

TLDR: We want to keep a stable ID for each launch so we can join CDT sessions.

Open source: no-op as perf metrics is only applicable for internal builds. See #16 

Please also see the internal RFC: https://www.internalfb.com/diff/D55164646

This launch ID is read from the query param; thus remains the same for each CDT refresh, either by pressing F5 or clicking "Reconnect DevTools":

https://github.com/facebookexperimental/rn-chrome-devtools-frontend/blob/9370b35e1979ec6953ff59ed78c8751e4f463750/front_end/ui/legacy/RemoteDebuggingTerminatedScreen.ts#L47

## Why

Since modern web standards relies on `visibilitychange` to signal the end of a session, we need to account for users showing and hiding the window multiple times. Closing the window is indistinguishable from hiding the window under this model.

* [MDN: Sending analytics at the end of a session](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon#sending_analytics_at_the_end_of_a_session)

This means each time the window is hidden, we treat this as one completed CDT session and submit end-of-session perf metrics as if the window is closed.

If the user was only hiding the window and came back later, we need a way to de-dup the subsequent events.

Therefore, we need to keep a stable ID across refreshes and `visibilitychange` events. For each launch, e.g. from iOS, Android, or Metro, a launch ID is generated in the internal version of Metro. [See internal diff](https://www.internalfb.com/diff/D55164646)

We'll then treat all events with the same launch ID as a single session in our metrics analytics.

Pending further discussion, we can also use this launch ID for Metro and CDP backend to attribute perf metrics outside of CDT.

# Stack

Stacked on #26

Depended by #28 

Preview this PR in isolation at https://github.com/EdmondChuiHW/rn-chrome-devtools-frontend2/pull/1/files

# Test plan

Non-null `launchId` in query param doesn't crash when launched from iOS, Android, and Metro (internal)

Null `launchId` query param: no-op (for open-source)

# Upstreaming plan

- [ ] This commit should be sent as a patch to the upstream `devtools-frontend` repo. I've reviewed the [contribution guide](https://docs.google.com/document/d/1WNF-KqRSzPLUUfZqQG5AFeU_Ll8TfWYcJasa_XGf7ro/edit#heading=h.9kj7femz1xg5).
- [x] This commit is React Native-specific and cannot be upstreamed.

### CODE REVIEW

File: front_end/core/host/RNPerfMetrics.ts [hunk 1]
```diff
@@ -5,26 +5,21 @@
...
 }
 
-class RNPerfMetricsImpl implements RNPerfMetrics {
+type UnsunscribeFn = () => void;
+class RNPerfMetrics {
   #listeners: Set<RNReliabilityEventListener> = new Set();
+  #launchId: string|null = null;
 
-  addEventListener(listener: RNReliabilityEventListener): () => void {
+  addEventListener(listener: RNReliabilityEventListener): UnsunscribeFn {
```
Comment (motiz88) [inline front_end/core/host/RNPerfMetrics.ts:22]: Nit: Let's fix the typo?
```suggestion
  addEventListener(listener: RNReliabilityEventListener): UnsubscribeFn {
```

File: front_end/entrypoints/rn_inspector/rn_inspector.ts [hunk 2]
```diff
@@ -22,6 +22,7 @@ import type * as InspectorBackend from '../../core/protocol_client/InspectorBack
 
 Host.RNPerfMetrics.registerPerfMetricsGlobalPostMessageHandler();
 
+Host.rnPerfMetrics.setLaunchId(Root.Runtime.Runtime.queryParam('launchId'));
```
Comment (huntie) [inline front_end/entrypoints/rn_inspector/rn_inspector.ts:25]: Different casing here compared to L23? 🤔

File: front_end/core/host/RNPerfMetrics.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ export function getInstance(): RNPerfMetrics {
 type UnsunscribeFn = () => void;
 class RNPerfMetrics {
   #listeners: Set<RNReliabilityEventListener> = new Set();
+  #launchId: string|null = null;
```
Comment (huntie) [inline front_end/core/host/RNPerfMetrics.ts:20]: nit / question: Do we have Prettier enabled on TS files?

File: front_end/entrypoints/rn_inspector/rn_inspector.ts (hunk 2 above)
Comment (EdmondChuiHW) [reply front_end/entrypoints/rn_inspector/rn_inspector.ts:25]: Yeah one is static/constructor. One is the global instance. Following the pattern of the upstream `UserMetrics`:

https://github.com/facebookexperimental/rn-chrome-devtools-frontend/blob/9370b35e1979ec6953ff59ed78c8751e4f463750/front_end/core/host/host-legacy.ts#L40-L48

File: front_end/core/host/RNPerfMetrics.ts (hunk 3 above)
Comment (EdmondChuiHW) [reply front_end/core/host/RNPerfMetrics.ts:20]: Yes. This is the formatted output, consistent with the upstream formatting, e.g.:

https://github.com/facebookexperimental/rn-chrome-devtools-frontend/blob/9370b35e1979ec6953ff59ed78c8751e4f463750/front_end/core/sdk/Cookie.ts#L19

The `npm` lint or `git cl format` commands are hit-or-miss tho. Tried running the linter in #12 which created divergent output vs upstream main.

The docs says the project format is `clang`:

https://docs.google.com/document/d/1WNF-KqRSzPLUUfZqQG5AFeU_Ll8TfWYcJasa_XGf7ro/view#heading=h.4lu5vvp3iz1w

https://chromium.googlesource.com/devtools/devtools-frontend/+/main/docs/README.md#useful-commands

So I'm running this formatter locally on save:

https://marketplace.visualstudio.com/items?itemName=xaver.clang-format

And my workspace-specific settings to avoid changing unmodified lines:
```json
// .vscode/settings.json
{
  …
  "editor.codeActionsOnSave": {
    "source.fixAll": true
  },
  "editor.formatOnSaveMode": "modificationsIfAvailable",
  "[typescript]": {
    "editor.defaultFormatter": "xaver.clang-format"
  }
  …
}
```

Comment (motiz88) [reply front_end/core/host/RNPerfMetrics.ts:20]: Clang does TypeScript?? what

Comment (EdmondChuiHW) [reply front_end/core/host/RNPerfMetrics.ts:20]: Yeah and Google is very proud of it in the doc above

> The JS code style of the DevTools is basically the Blink Coding Style Guidelines, which are for C++, but we'll apply them to JS. (Yup)

**Review by huntie — APPROVED** [review]
LGTM :)
